### PR TITLE
feat: Velodrome V2 pool module

### PR DESCRIPTION
**Pull request by mikeyrf**
## Description

Fixes # (issue)

## Checklist

- [ ] New Contracts have been tested
- [ ] Lint has been run
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: test/router/modules/pool/LinkedPoolVelodromeV2ModuleOptUSDC.t.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    function addPools() public override {
+        addPool({
+            poolName: "USDC/USDT",
+            nodeIndex: 0,
+            pool: VEL_V2_USDCUSDT_POOL,
+            poolModule: address(velodromeV2Module)
+        });
+        addPool({poolName: "nUSD/USDC", nodeIndex: 0, pool: NUSD_POOL}); // Q: why is this node index 0?
```
Comment (mikeyrf) [inline test/router/modules/pool/LinkedPoolVelodromeV2ModuleOptUSDC.t.sol:54]: @ChiTimesChi question for you here as I'm not understanding exactly how node index specification works for this function

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        router = IVelodromeV2Router(_router);
+    }
+
+    function poolSwap(
+        address pool,
+        IndexedToken memory tokenFrom,
+        IndexedToken memory tokenTo,
+        uint256 amountIn
+    ) external returns (uint256 amountOut) {
+        // TODO: sanity checks on tokenFrom, tokenTo indices <= 1?
```
Comment (ChiTimesChi) [inline contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:28]: The caller of `poolSwap()` is responsible for passing valid `IndexedToken` entries (e.g. the ones where token actually matches the provided index). Given that we only need the token addresses here, I think we're fine not checking the indexes.

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        IVelodromeV2Router.Route[] memory routes = new IVelodromeV2Router.Route[](1);
+        routes[0] = IVelodromeV2Router.Route({
+            from: tokenFrom.token,
+            to: tokenTo.token,
+            stable: stable,
+            factory: factory
+        });
+
+        uint256[] memory amounts = new uint256[](2);
+        amounts = router.swapExactTokensForTokens(amountIn, 0, routes, address(this), block.timestamp);
```
Comment (ChiTimesChi) [inline contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:43]: Was the implementation where Module directly interacts with the pool (vs interacting with Velodrome Router) considered?

For instance, in #246 the Router interaction was chosen, as the direct interaction with UniV3 pool requires implementing a callback, which is not possible for a Module that gets delegate calls.

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    /// @dev Careful with pool.getAmountOut as will return token1 as in if give invalid tokenIn address (token not in pool)
+    function getPoolQuote(
+        address pool,
+        IndexedToken memory tokenFrom,
+        IndexedToken memory tokenTo,
+        uint256 amountIn,
+        bool probePaused
+    ) external view returns (uint256 amountOut) {
+        address tokenIn = tokenFrom.token;
```
Comment (ChiTimesChi) [inline contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:55]: I think checking `(tokenFrom, tokenTo)` against `pool.tokens()` makes sense here. Indexes don't really need to be checked, as they are not going to be later used anyway.

File: test/router/modules/pool/LinkedPoolVelodromeV2ModuleOptUSDC.t.sol (hunk 1 above)
Comment (ChiTimesChi) [reply test/router/modules/pool/LinkedPoolVelodromeV2ModuleOptUSDC.t.sol:54]: In this scenario you're attaching two pools to `USDC` node (which is the root token in example):
- VelodromeV2 USDC/USDT (attached to USDC node with index 0)
- DefaultPool nUSD/USDC (attached to USDC node with index 0)

Which results in the following structure:
image

In production, however, the root token would be nUSD, and the setup would be different:
- DefaultPool nUSD/USDC (attached to nUSD node with index 0)
- VelodromeV2 USDC/USDT (attached to USDC node with index 1)

Which would result in the following structure:
image

Comment (ChiTimesChi) [reply test/router/modules/pool/LinkedPoolVelodromeV2ModuleOptUSDC.t.sol:54]: As mentioned in #245, the ideal test here would be one that mimics the upcoming production as much as possible, so changing the expected tokens order and order of pools addition should be good practice.

**Review by ChiTimesChi — CHANGES_REQUESTED** [review]
Good stuff, but needs some more polishing before prod. Requesting a few changes below.

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol (hunk 3 above)
Comment (mikeyrf) [reply contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:43]: Can do either way. Was being consistent with Uni v3 approach, but now I understand why you took the route you took. Will fix to go through pool to limit extra calls.

**Comment by mikeyrf** [comment]
@ChiTimesChi fixes made given the PR review. Should be good to go now, but lemme know if otherwise

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        IndexedToken memory tokenFrom,
+        IndexedToken memory tokenTo,
+        uint256 amountIn,
+        bool probePaused
+    ) external view returns (uint256 amountOut) {
+        (address token0, address token1) = IVelodromeV2Pool(pool).tokens();
+        address tokenIn = tokenFrom.token;
+        address tokenOut = tokenTo.token;
+        require(tokenIn == token0 || tokenIn == token1, "tokenFrom not token in pool");
+        require(tokenOut == token0 || tokenOut == token1, "tokenTo not token in pool");
```
Comment (ChiTimesChi) [inline contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:61]: This will not catch the case where `tokenIn == tokenOut == token0`.

Just checking if the provided tokens are `(token0, token1)` or `(token1, token0)` should do the trick:
```solidity
require(tokenIn == token0 && tokenOut == token1 || tokenIn == token1 && tokenOut == token0, "...");
```

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+contract VelodromeV2Module is OnlyDelegateCall, IPoolModule {
+    using SafeERC20 for IERC20;
+
+    IVelodromeV2Router public immutable router;
+
+    constructor(address _router) {
+        router = IVelodromeV2Router(_router);
+    }
+
+    function poolSwap(
```
Comment (ChiTimesChi) [inline contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:24]: This (and other functions) needs to inherit the documentation:
```solidity
/// @inheritdoc IPoolModule
```

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            stable: stable,
+            factory: factory
+        });
+
+        uint256[] memory amounts = new uint256[](2);
+        amounts = router.swapExactTokensForTokens(amountIn, 0, routes, address(this), block.timestamp);
+        amountOut = amounts[1];
+    }
+
+    /// @dev Careful with pool.getAmountOut as will return token1 as in if give invalid tokenIn address (token not in pool)
```
Comment (ChiTimesChi) [inline contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:49]: This is outdated now.

File: test/router/modules/pool/velodrome/VelodromeV2ModuleOpt.t.sol [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    function testPoolSwapRevertsWhenDirectCall() public {
+        vm.expectRevert("Not a delegate call");
+        uniswapV3Module.poolSwap({
+            pool: UNI_V3_USDC_POOL,
+            tokenFrom: IndexedToken({index: 0, token: USDC}),
+            tokenTo: IndexedToken({index: 1, token: USDC_E}),
+            amountIn: 100 * 10**6
+        });
+    }
+    */
```
Comment (ChiTimesChi) [inline test/router/modules/pool/velodrome/VelodromeV2ModuleOpt.t.sol:117]: Time to uncomment the test :)

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol (hunk 5 above)
Comment (mikeyrf) [reply contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:61]: Good catch sorry. Will fix.

**Comment by mikeyrf** [comment]
Fixes added as well

File: contracts/router/modules/pool/velodrome/VelodromeV2Module.sol [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        uint256 amountIn,
+        bool probePaused
+    ) external view returns (uint256 amountOut) {
+        (address token0, address token1) = IVelodromeV2Pool(pool).tokens();
+        address tokenIn = tokenFrom.token;
+        address tokenOut = tokenTo.token;
+        require(
+            (tokenIn == token0 && tokenOut == token1) || (tokenIn == token1 && tokenOut == token0),
+            "tokens not in pool"
+        );
```
Comment (ChiTimesChi) [inline contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:64]: Even though we're not tracking coverage %, would be worth to add tests covering these edge cases

Comment (mikeyrf) [reply contracts/router/modules/pool/velodrome/VelodromeV2Module.sol:64]: @ChiTimesChi test added in [latest commit](https://github.com/synapsecns/synapse-contracts/pull/251/commits/fa56e3aa9f0f7057b50273f7ce11aa1437c0b6af).

**Review by ChiTimesChi — APPROVED** [review]
LGTM
